File: src/main/java/net/onelitefeather/stardust/tasks/UserTask.java
package net.onelitefeather.stardust.tasks;

import net.kyori.adventure.text.minimessage.MiniMessage;
import net.onelitefeather.stardust.FeatherEssentials;
import net.onelitefeather.stardust.api.user.IUser;

public record UserTask(FeatherEssentials featherEssentials) implements Runnable {
    @Override
    public void run() {
        this.featherEssentials.getServer().getOnlinePlayers().forEach(player -> {
            IUser user = this.featherEssentials.getUserManager().getUser(player.getUniqueId());
            if(user != null) {
                if(user.isVanished()) {
                    player.sendActionBar(MiniMessage.miniMessage().deserialize(this.featherEssentials.getMessage("plugin.vanish-actionbar")));
                }
            }
        });
    }
}

File: src/main/kotlin/net/onelitefeather/stardust/tasks/UserTask.kt
package net.onelitefeather.stardust.tasks

import net.kyori.adventure.text.Component
import net.onelitefeather.stardust.StardustPlugin
import net.onelitefeather.stardust.util.PlayerUtils
import java.util.logging.Level

class UserTask(val stardustPlugin: StardustPlugin) : Runnable, PlayerUtils {

    override fun run() {
        try {

            stardustPlugin.server.onlinePlayers.forEach { player ->

                val user = stardustPlugin.userService.getUser(player.uniqueId) ?: return
                if (user.isVanished()) {
                    player.sendActionBar(Component.translatable("plugin.vanish-actionbar"))

                    //Keep fly mode if the player's gamemode was changed to survival or adventure
                    if(!player.allowFlight && !canEnterFlyMode(player)) {
                        player.allowFlight = true
                    }

                    return
                }

                if(!canEnterFlyMode(player)) {
                    player.allowFlight = user.isFlying()
                }
            }
        } catch (e: Exception) {
            stardustPlugin.logger.log(Level.SEVERE, "Could not run for loop", e)
        }
    }
}